[C++] Refactor to conform to Google C++ style guide

dbaileychess opened the pull request:
This refactor conforms all the C++ code to follow the [Google C++ style guide](https://google.github.io/styleguide/cppguide.html) which is the requested [style of flatbuffers](https://github.com/google/flatbuffers/blob/master/CONTRIBUTING.md). 

This change was done by automated code formatting tools and introduces no logical changes.

The majority of the changes are:

- Line limits to [80 characters](https://google.github.io/styleguide/cppguide.html#Line_Length)
- [Horizontal spacing](https://google.github.io/styleguide/cppguide.html#Horizontal_Whitespace)
- Proper indentation [2 space](https://google.github.io/styleguide/cppguide.html#Spaces_vs._Tabs)
- Ordering and spacing of [includes](https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes)
- [Expression consistency](https://google.github.io/styleguide/cppguide.html#Pointer_and_Reference_Expressions)
   - The asterisk and ampersand symbols are all aligned to the variable name. Though alignment to the type is OK as well.

Tested: `flattest` ran successfully and `test/generate_code.sh` produced no changes to generated files.

I grouped the refactors into separate commits, in case we want to merge them separately. I can squash them if required.

aardappel commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -658,7 +657,7 @@ class DetachedBuffer {
         size_(sz) {}
 
-  // clang-format off
+// clang-format off
```
these `// clang-format off` indentation changes are unfortunate. The whole reason we have them is because clang-format doesn't support indented #ifdefs, and indenting them makes code more "scannable" because they don't interrupt the indentation of the code. This once more interrupts the regular indentation of the code.

I am not sure why it is happening, surely it knows how to indent `//` correctly? Is this a special case for `// clang-format off` ??

aardappel replied:
Or maybe it now thinks that since its followed by a line with `#`, it must be in the first column? Maybe we (gasp) need a blank line between them? :(

Note in https://clang.llvm.org/docs/ClangFormatStyleOptions.html IndentPPDirectives there is `BeforeHash` we want, but supposedly it doesn't work yet.. https://bugs.llvm.org/show_bug.cgi?id=36019 https://bugs.llvm.org/show_bug.cgi?id=36020

dbaileychess replied:
Appears the issue is a [bug in clang-format](https://bugs.llvm.org/show_bug.cgi?id=39280) with improper formatting of the "// clang-format off" line when it precedes a #. And yes, adding a blank line afterwards avoids the issue. And I noticed there are a bunch of instances already in the code with a blank line in that position.

A simple fix is to run clang-format twice, as it fixes it self!

aardappel commented:
Thanks! This is generally very welcome, cleaning up a lot of the code from recent PRs that were not clang-formatted, but there is a bit of a regression in how `// clang-format off` is handled.. this was not a problem before. What version of clang-format is this?

dbaileychess commented:
Fixed all the issues with the improper formatting of the clang-format lines. I also noticed that the Kotlin generator never closed the clang-format off block, so I added the closing block. This then caused the whole idl_gen_kotlin.cpp file to be re-written since it was way out of the correct style.

aardappel commented:
Awesome, thanks :)

aardappel commented:
I was thinking we should have an automated "github actions" or something that forces a clang-format on each PR.. but given that current bug in clang-format, that may be more pain then helpful.

In the mean time, I guess we'll have these PRs on occasion :)

dbaileychess commented:
You could include instructions to run clang-format before submitting a PR, similar to running `tests/generate_code.sh`

aardappel commented:
The problem is if they are just instructions, then most people will end up not doing it, and then anyone who is following those instructions inevitably ends up with formats from other people.

Unless we just ask people to `git clang-format HEAD^`, which would mostly also be sufficient.

I get the idea that most people don't read the PR instructions though..

aardappel commented:
Ok, maybe this helps: https://github.com/google/flatbuffers/commit/33d5dd9bdd99baf7a0005be36ca468977fdef46d
